### PR TITLE
updates FMS to 2020.04.01

### LINKED ISSUE #391: update FMS to 2020.04.01
## Description
The current UFS Weather Model uses FMS 2019.01.03 which uses the old set of mpp_io subroutines and functions.  In newer versions of FMS, fms2_io has been implemented which improves upon mpp_io.

## Solution
FMS 2020.04.01 that defaults to fms2_io is now available to be integrated into the UFS Weather Model.

### PR DESCRIPTION
**Pull request by mlee03**
## Description

This PR updates FMS to version ``2020.04.01`` and includes two notable changes:

1.  FMS 2020.04.01 will default to use the newer io implementation called fms2_io.  To use the old mpp_io that is in 2019.01.03, the namelist variable ``use_mpp_io=.true.`` is required for  interpolator_nml, xgrid_nml, amip_interp_nml,  and diag_manager_nml;  and ``use_mpp_bug=.true.`` in data_override_nml

2. An additional option has been implemented in the file section of the diag_table and only affects files using the optional diag_table features.  Users can now specify the last field as "begin", "middle", or "end" in order to have more control of the output filename.  An example specification is shown below:  
<pre>"ocn%4yr%2mo%2dy%2hr", 6,  "hours", 1, "hours", "time", 6, "hours", "1901 1 1 0 0 0", 0, "", "begin"</pre>
Given the above specification, the file corresponding to the first 6 hours of the run will be named 
``ocn_2016_10_03_00.nc`` for "begin"
``ocn_2016_10_03_03.nc`` for "middle"
 ``ocn_2016_10_03_06.nc`` for "end"
If the last field is not specified , "middle" will be used as default.  More details are provided [here](https://github.com/NOAA-GFDL/FMS/pull/660#issue-552786503)

CMakeLists.txt and fms_cmake.files have been modified to take into account files that have been deleted and added.  Two metadata differences may be observed with nccmp: 

1. NetCDF file format change from NC_FORMAT_NETCDF4_CLASSIC to NC_FORMAT_64BIT.  Users can  specify ``netcdf_default_format="classic"`` in ``fms2_io_nml`` to use NC_FORMAT_NETCDF4_CLASSIC.
2. Checksum differences in FV3 32-bit runs due to a bug fix in mpp_chksum.

### Issue(s) addressed
- fixes #391

## Testing
All regression tests pass on Orion with Intel/2018.4 

## Dependencies
There are no dependencies

### CODE REVIEW

**Comment by aerorahul** [comment]
@mlee03 
This version of FMS has CMake build functionality.
We should be using that instead of maintaining our own version of `fms_files.cmake` and other unnecessary code in the parent `CMakeLists.txt`.

Lines 134-170 in the top-level `CMakeLists.txt` can be replaced by
```
add_subdirectory(FMS)
```

File: fms_files.cmake [hunk 1]
```diff
@@ -28,57 +30,61 @@ list(APPEND fms_src_files
...
+    FMS/fms2_io/fms_netcdf_unstructured_domain_io.F90
+    FMS/fms2_io/fms2_io.F90
+    FMS/fms2_io/netcdf_io.F90
     FMS/horiz_interp/horiz_interp.F90
     FMS/horiz_interp/horiz_interp_bicubic.F90
     FMS/horiz_interp/horiz_interp_bilinear.F90
     FMS/horiz_interp/horiz_interp_conserve.F90
     FMS/horiz_interp/horiz_interp_spherical.F90
     FMS/horiz_interp/horiz_interp_type.F90
+    FMS/horiz_interp/horiz_interp.F90
```
Comment (junwang-noaa) [inline fms_files.cmake]: is the line same as line 43?

Comment (aerorahul) [reply fms_files.cmake]: @junwang-noaa This file should not exist, if we move to this release of FMS.

File: fms_files.cmake [hunk 2]
```diff
@@ -28,57 +30,61 @@ list(APPEND fms_src_files
...
     FMS/time_manager/get_cal_time.F90
     FMS/time_manager/time_manager.F90
     FMS/topography/gaussian_topog.F90
     FMS/topography/topography.F90
     FMS/tracer_manager/tracer_manager.F90
     FMS/tridiagonal/tridiagonal.F90
 
-    FMS/memutils/memuse.c
+    FMS/affinity/affinity.c
+    FMS/mpp/mpp_memuse.c
```
Comment (junwang-noaa) [inline fms_files.cmake:81]: Is there any documentation on what mpp_memuse can do?

**Comment by mlee03** [comment]
@aerorahul, I will make the change and test it.

**Comment by aerorahul** [comment]
> @aerorahul, I will make the change and test it.

Happy to help.  Let me know.

**Comment by junwang-noaa** [comment]
I want to confirm that 1) the default io is now fms2_io and there is no results change when using nccmp to compare data itself , but meta are changed. 2) No change is required in MOM6 and fv3dycore for this PR.  3) No new baseline is required for this PR.

**Comment by mlee03** [comment]
@junwang-noaa  
1) The default io is fms2_io.  There are no data differences with nccmp, only metadata differences.
2) No change is required in MOM6 and fv3dycore for this PR, and 
3) No new baseline is required for this PR.

I'm also working on getting the answer about mpp_memuse.c documentation.

**Comment by bensonr** [comment]
@mlee03 @junwang-noaa - although the default is fms2_io, the fv3dycore does not yet utilize it for restart IO - only diagnostics within FMS.   fms2_io for fv3dycore restarts is in the pipeline and will be released when fully tested.  The IO from within [FV3GFS_io.F90](https://github.com/NOAA-EMC/fv3atm/blob/develop/io/FV3GFS_io.F90) will also need to be inspected to ensure it is using fms2_io.

**Comment by mlee03** [comment]
@junwang-noaa,  do I need to update the .gitmodules file to point FMS to 2020.04.01?

**Comment by junwang-noaa** [comment]
Yes.

**Comment by mlee03** [comment]
@junwang-noaa, done.

**Comment by climbfuji** [comment]
When I compile this code on Cheyenne with GNU, I see the following warning:
```
/glade/work/heinzell/fv3/ufs-weather-model/ufs-weather-model-fms-2020-04-01/gnu/FMS/fms2_io/netcdf_io.F90:25:17:

   25 | use, intrinsic :: iso_fortran_env
      |                 1
Warning: Use of the NUMERIC_STORAGE_SIZE named constant from intrinsic module ISO_FORTRAN_ENV at (1) is incompatible with option -fdefault-real-8
```
I don't think this is related to the crashes of all regression tests when compiled with GNU in REPRO or PROD mode (it works in DEBUG mode), but in any case somebody should look at this.

**Comment by mlee03** [comment]
@climbfuji thank you for bringing this up.  We will look into it.

**Review by climbfuji — COMMENTED** [review]
Please delete file `RegressionTests_cheyenne.gnu.log` in your top-level ufs-weather-model directory, this is a duplicate in the wrong place. The file is also updated in `tests` later (which is the correct place).

**Comment by climbfuji** [comment]
> Please delete file `RegressionTests_cheyenne.gnu.log` in your top-level ufs-weather-model directory, this is a duplicate in the wrong place. The file is also updated in `tests` later (which is the correct place).

I see, you just did. Thanks! I'll send you a new cheyenne.gnu log once the rerun of `rt.sh` is done.

**Review by climbfuji — APPROVED** [review]

**Review by junwang-noaa — APPROVED** [review]

**Pull request merged**